File: Java/stringmatching.java
import java.io.*;
import java.util.*;

public class stringmatching {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        StringBuilder ans = new StringBuilder();
        while (sc.hasNextLine()) {
            String b = sc.nextLine();
            String a = sc.nextLine();
            int[] lps = getLPS(b);
            int i = 0, j = 0;
            while (i < a.length()) {
                if (a.charAt(i) == b.charAt(j)) {
                    i++;
                    j++;
                }
                if (j == b.length()) {
                    ans.append(i - j).append(' ');
                    j = lps[j - 1];
                } else if (i < a.length() && a.charAt(i) != b.charAt(j)) {
                    if (j > 0)
                        j = lps[j - 1];
                    else
                        i++;
                }
            }
            ans.append('\n');
        }
        System.out.println(ans);
    }

    static int[] getLPS(String a) {
        int[] arr = new int[a.length()];
        int x = 1, len = 0;
        while (x < a.length()) {
            if (a.charAt(x) == a.charAt(len)) {
                arr[x++] = ++len;
            } else {
                if (len > 0)
                    len = arr[len - 1];
                else
                    arr[x++] = len;
            }
        }
        return arr;
    }
}

File: C++/stringmatching.cpp
#include <bits/stdc++.h>

using namespace std;

void getLPS(int arr[], int size, string a) {
    arr[0] = 0;
    int x = 1, len = 0;
    while (x < size) {
        if (a[x] == a[len]) {
            arr[x++] = ++len;
        } else {
            if (len)
                len = arr[len - 1];
            else
                arr[x++] = len;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    while (cin) {
        string a, b;
        getline(cin, b);
        getline(cin, a);
        int lps[b.size()];
        getLPS(lps, b.size(), b);
        int i = 0, j = 0;
        while (i < a.size()) {
            if (a[i] == b[j])
                i++, j++;
            if (j == b.size()) {
                cout << i - j << ' ';
                j = lps[j - 1];
            } else if (i < a.size() && a[i] != b[j]) {
                if (j)
                    j = lps[j - 1];
                else
                    i++;
            }
        }
        cout << '\n';
    }
    return 0;
}